### PR TITLE
KAFKA-17469: Moved share external interfaces to share module.

### PR DESCRIPTION
**Pull request by smjn**
### What
- Moved share group/coordinator external interface classes to share module. 

### Why
- A number of common DTO classes and interfaces exposed for KIP-932 and related to share group/coordinator persistence are currently housed in the `server-common` module.
- Post introduction of the above classes, a new module `share` was introduced to encapsulate entities which could be required by various other modules (`group-coordinator`, `core`, `share-coordinator`),  basically external interface.
- Hence, `server-common` is no longer the ideal location for external share entities and is more suited for very general code.
- Therefore, in this PR we have moved the classes from `server-common` to `share` and removed the applicable package from server-common.
- As a result of this refactor, some other classes in various modules were updated to reflect the package name change along with import control and `build.gradle`.

### Testing
No business logic change, clean build locally.

### CODE REVIEW

**Review by apoorvmittal10 — APPROVED** [review]
LGTM!

File: share/src/main/java/org/apache/kafka/server/share/DeleteShareGroupStateParameters.java [hunk 1]
```diff
@@ -15,7 +15,7 @@
  * limitations under the License.
  */
 
-package org.apache.kafka.server.group.share;
+package org.apache.kafka.server.share;
```
Comment (apoorvmittal10) [inline share/src/main/java/org/apache/kafka/server/share/DeleteShareGroupStateParameters.java:18]: Should we structure the classes in `group` package inside `share` module itself?

Comment (smjn) [reply share/src/main/java/org/apache/kafka/server/share/DeleteShareGroupStateParameters.java:18]: These have little to do with share group and more with share coordinator.
Additionally we will not have any other groups in share module so would not add much.

Comment (apoorvmittal10) [reply share/src/main/java/org/apache/kafka/server/share/DeleteShareGroupStateParameters.java:18]: Then should it be in `coordinator` package in `share`? I am just trying to get it better structured rather pushing everything in parent package.

Comment (smjn) [reply share/src/main/java/org/apache/kafka/server/share/DeleteShareGroupStateParameters.java:18]: For the callers the main entry point is NoOpStatePersister/Persister/DefaultStatePersister. 
All the other classes are wrappers for the actual RPC records. 

Ideally callers should not be aware that there is even a coordinator. They will however need to create objects to pass parameters into and fetch results from these interfaces.

The thinking behind the current structure was not to expose to the callers that there is a specific handler/share coordinator which these interfaces delegate the call to.

We followed similar approach in coordinator-common where we just kept one package for Records/metrics/runtime etc.

**Review by mumrah — APPROVED** [review]
LGTM, thanks @smjn!

File: share/src/main/java/org/apache/kafka/server/share/DeleteShareGroupStateParameters.java (hunk 1 above)
Comment (apoorvmittal10) [reply share/src/main/java/org/apache/kafka/server/share/DeleteShareGroupStateParameters.java:18]: Ok, not fully convinced though but we can see if required specific sub package later.

**Pull request merged**